src/orion/client/runner.py

```diff
@@ -0,0 +1,339 @@
+            results = self.client.executor.async_get(
+                self.futures, timeout=self.gather_timeout
+            )
+        except KeyboardInterrupt:
+            self._release_all()
+            raise
+
+        to_be_raised = None
+        log.debug(f"Gathered new results {len(results)}")
+
+        # register the results
+        for result in results:
+            trial = self.pending_trials.pop(result.future)
+
+            if isinstance(result, AsyncResult):
+                try:
+                    # NB: observe release the trial already
+                    self.client.observe(trial, result.value)
+                    self.trials += 1
+                except InvalidResult as exception:
+                    # stop the optimization process if we received `InvalidResult`
+                    # as all the trials are assumed to be returning those
+                    to_be_raised = exception
+                    self.client.release(trial, status="broken")
+
+            if isinstance(result, AsyncException):
+                if (
+                    isinstance(result.exception, ExecutionError)
+                    and result.exception.return_code == self.interrupt_signal_code
+                ):
+                    to_be_raised = KeyboardInterrupt()
+                    self.client.release(trial, status="interrupted")
+                    continue
+
+                # Regular exception, might be caused by the choosen hyperparameters
+                # themselves rather than the code in particular (like Out of Memory error
+                # for big batch sizes)
+                exception = result.exception
+                self.worker_broken_trials += 1
+                self.client.release(trial, status="broken")
+
+                if self.on_error is None or self.on_error(
+                    self, trial, exception, self.worker_broken_trials
+                ):
+                    log.error(result.traceback)
+
+                else:
+                    log.error(str(exception))
+                    log.debug(result.traceback)
+
+        # if we receive too many broken trials, it might indicate the user script
+        # is broken, stop the experiment and let the user investigate
+        if self.is_broken:
+            to_be_raised = BrokenExperiment(
+                "Worker has reached broken trials threshold"
+            )
+
+        if to_be_raised is not None:
+            log.debug("Runner was interrupted")
+            self._release_all()
+            raise to_be_raised
+
+        return len(results)
+
+    def _release_all(self):
+        """Release all the trials that were reserved by this runner.
+        This is only called during exception handling to avoid retaining trials
+        that cannot be retrieved anymore
+
+        """
+        # Sanity check
+        for _, trial in self.pending_trials.items():
+            try:
+                self.client.release(trial)
+            except AlreadyReleased:
```

Comment:
Not tested

Reply:
Done here https://github.com/Epistimio/orion/pull/684/commits/0d390c553d734d53440bff566f85eb124b30152e